docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,378 @@
+    storage_token: Vec<u8>, // (??) not clear on the type here, but whatever.
+
+    // Info about the collections managed by this engine. In most cases there
+    // will be exactly 1.
+    collections: HashMap<String, CollectionState>
+}
+
+// The state for a collection. Most engines will manage a single collection, but
+// there may be many.
+struct CollectionState {
+    // The info/collections response for this collection.
+    collection_info: CollectionInfo,
+
+    // The current syncID for the engine, as read from meta/global. Will never
+    // be None as the sync manager is responsbile for generating an ID if
+    // it doesn't exist.
+    // Engines are responsible for persisting this and taking action when it
+    // differs from last time.
+    sync_id: SyncGuid,
+
+    // The current version for the engine, as read from meta/global.
+    // Not clear that "u32" is the best option here - I guess it's OK so
+    // long as we treat an invalid u32 representation sanely.
+    // Note that persisting this may not be required if there's only one
+    // version supported, as the value can be compared against a hard-coded
+    // value in the engine.
+    version: u32,
+
+    // The keys to be used for this collection. This may be the default key (ie,
+    // used for all collections) or one specific to the collection, but the
+    // engine doesn't need to know that.
+    keys: KeyBundle,
+
+    // Commands the engine should act upon before syncing.
+    commands: Vec<Command>
+}
+
+// Used to support commands. Will either be a strongly typed enum, or a
+// weakly typed thing (such as {command: String, args: Vec<String>})
+// TBD.
+struct Command {
+    ...
+}
+```
+
+## APIs
+## Sync Manager
+
+```rust
+// Get the list of engines which are declined. There's a bit of confusion here
+// between an "engine" and a "collection" - eg, assuming history ends up using
+// 2 collections, the UI will want 1 entry.
+// Maybe we define this as a "collection" and the app should ignore collections
+// it doesn't understand?
+// Will return None if we haven't yet done a first sync so don't know. Will
+// return an empty vec if we have synced but no engines are declined.
+fn get_declined() -> Option<Vec<String>>;
+
+// Ditto here - a bit of confusion between "collection" and "engine" here.
+fn enable_engine(String, bool) -> Result<()>;
+
+// TODO: let's think a little more about how desktop uses a single "addons"
+// pref which controls addons *and* storage.sync, and how a single "Forms"
+// pref controls 2 discrete engines.
+
+// Do a sync now. There's a special error code to indicate that something
+// seems to have gone wrong with auth, which is surfaced to the app, so
+// it can call on FxA to do magic.
+fn sync_now(engines: Option<Vec<String>>) -> Result<()>;
+
+// Stop syncing ASAP. This might be called as the app is shutting down or when
+// the app has lost wifi, for example.
+fn stop();
```

Comment:
I don't see something like Fenix using it in practice, as sync operations will be executed in background workers that survive app shutdowns.